### PR TITLE
[pallas backend] implement gpu tiles/mask for power of 2

### PR DESCRIPTION
**Pull request by oulgen**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* __->__ #167584

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

### CODE REVIEW

File: torch/_inductor/codegen/pallas.py [hunk 1]
```diff
@@ -529,16 +530,65 @@ def codegen_kernel(self, name: Optional[str] = None) -> str:  # type: ignore[ove
...
+        code.writeline(
+            f"def {jit_wrapper_name}(out_shape, out_dtype, device_type, *kernel_refs):"
+        )
         with code.indent():
-            code.writeline("out_spec = jax.ShapeDtypeStruct(out_shape, out_dtype)")
-            code.writeline("return pl.pallas_call(")
+            code.writeline(
+                "# On GPU (Triton backend), Pallas requires each dimension to be power of 2"
+            )
+            code.writeline("if device_type == 'cuda':")
```
Comment (jansel) [inline torch/_inductor/codegen/pallas.py]: We should know the device at compile time, no need for this to be a runtime check.

File: torch/_inductor/codegen/pallas.py [hunk 2]
```diff
@@ -529,16 +530,65 @@ def codegen_kernel(self, name: Optional[str] = None) -> str:  # type: ignore[ove
...
+            with code.indent():
+                code.writeline("# Pad each dimension of each input to power of 2")
+                code.writeline("kernel_refs_padded = []")
+                code.writeline("for ref in kernel_refs:")
+                with code.indent():
+                    code.writeline(
+                        "padded_shape = tuple(next_power_of_2(d) for d in ref.shape)"
```
Comment (jansel) [inline torch/_inductor/codegen/pallas.py]: Wait, do we need the inputs to be a power of 2 or the tile sizes?  Triton only requires tile sizes, padding inputs would be very bad.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/pallas.py (hunk 2 above)
Comment (oulgen) [reply torch/_inductor/codegen/pallas.py]: oops sorry, i meant to change this PR. I need to instead
* pass in a block size to the kernel (currently we are not passing anything so i assume we are using some default)
* mask out extra bits

@jansel I assume this tech already exists for triton backend, so i can probably use that?

Comment (jansel) [reply torch/_inductor/codegen/pallas.py]: In the Triton backend we defer block size selection until runtime in the runtime wrappers (with autotuning to help select it).

File: torch/_inductor/codegen/pallas.py [hunk 3]
```diff
@@ -404,14 +415,98 @@ def _get_index_expr(self, index: sympy.Expr) -> tuple[str, bool]:
         else:
             return self._get_index_str(index), False
 
+    def _determine_masked_ops_for_kernel(self) -> bool:
+        """
+        Determine if we should use masked ops for this entire kernel.
+
+        Masked ops with pl.ds(block_size) flatten tensors to 1D, which works when:
+        1. We're on GPU (CUDA backend uses Triton which requires power-of-2 sizes)
+        2. All tensors are already 1D (so flattening doesn't change dimensionality)
```
Comment (oulgen) [inline torch/_inductor/codegen/pallas.py:424]: @jansel i tried to add support for 2d broadcast with mask but BlockSpec is a little hard to work with, so left that out for now

**Review by jansel — APPROVED** [review]
